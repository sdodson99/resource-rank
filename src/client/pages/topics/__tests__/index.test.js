import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { createRenderer } from 'react-test-renderer/shallow';
import Topics from '../index.page';
import { useRouter } from 'next/router';
import useAuthenticationState from '@/hooks/use-authentication-context';
import useTopicSearch from '@/hooks/topics/use-topic-search';

jest.mock('next/router');
jest.mock('@/hooks/use-authentication-context');
jest.mock('@/hooks/topics/use-topic-search');

describe('<Topics />', () => {
  let searchQuery;
  let mockProcessSearch;

  beforeEach(() => {
    searchQuery = 'search';
    useRouter.mockReturnValue({
      query: {
        q: searchQuery,
      },
    });

    useAuthenticationState.mockReturnValue({ isLoggedIn: false });

    mockProcessSearch = jest.fn();
    useTopicSearch.mockReturnValue({
      data: {
        topics: {
          items: [
            {
              id: '1',
              name: 'name',
              slug: 'slug',
            },
            {
              id: '2',
              name: 'name',
            },
          ],
        },
      },
      error: null,
      isLoading: false,
      search: '',
      currentSearch: '',
      processSearch: mockProcessSearch,
    });
  });

  afterEach(() => {
    useRouter.mockReset();
    useAuthenticationState.mockReset();
    useTopicSearch.mockReset();
  });

  it('should mount', () => {
    render(<Topics />);

    const page = screen.getByTestId('Topics');

    expect(page).toBeInTheDocument();
  });

  it('should render correctly', () => {
    const page = createRenderer().render(<Topics />);

    expect(page).toMatchSnapshot();
  });

  it('should process search on search input', () => {
    const search = '123';
    render(<Topics />);
    const searchInput = screen.getByTestId('SearchInput');

    fireEvent.input(searchInput, {
      target: {
        value: search,
      },
    });

    expect(mockProcessSearch).toBeCalledWith(search);
  });
});
