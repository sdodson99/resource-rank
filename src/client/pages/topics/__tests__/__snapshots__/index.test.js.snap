// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Topics /> should render correctly 1`] = `
<div
  data-testid="Topics"
>
  <BreadcrumbLayout
    breadcrumbs={
      Array [
        Object {
          "title": "Topics",
          "to": "/topics",
        },
      ]
    }
  >
    <NextSeo
      openGraph={
        Object {
          "description": "Find the topic you are planning to learn, or find a new topic to learn!",
          "title": "Browse Topics - Resource Rank",
        }
      }
      title="Topics"
    />
    <PageHeaderButton
      buttonContent="New"
      hideButton={true}
      linkTo="/topics/new"
      title="Topics"
      titleClassName="text-4xl"
    />
    <div
      className="mt-8 flex flex-col"
    >
      <input
        className="flex-grow form-control-lg"
        data-testid="SearchInput"
        onChange={[Function]}
        placeholder="Search topics..."
        type="text"
        value="search"
      />
      <div
        className="mt-8"
      >
        <LoadingErrorEmptyDataLayout
          dataDisplay={
            <TopicListing
              onPageClick={[MockFunction]}
              pageCount={NaN}
              selectedPage={1}
              topics={
                Array [
                  Object {
                    "id": "1",
                    "name": "name",
                    "slug": "slug",
                  },
                  Object {
                    "id": "2",
                    "name": "name",
                  },
                ]
              }
            />
          }
          errorDisplay={
            <div
              className="text-center sm:text-left error-text"
            >
              Failed to load topics.
            </div>
          }
          hasData={true}
          hasError={false}
          isLoading={false}
          loadingDisplay={
            <div
              className="text-center"
            >
              <LoadingSpinner
                color="#292929"
                size={50}
              />
            </div>
          }
          noDataDisplay={
            <div
              className="text-center sm:text-left"
            >
              No topics matching 'currentSearch' have been created.
            </div>
          }
        />
      </div>
    </div>
  </BreadcrumbLayout>
</div>
`;

exports[`<Topics /> should render correctly with no current search 1`] = `
<div
  data-testid="Topics"
>
  <BreadcrumbLayout
    breadcrumbs={
      Array [
        Object {
          "title": "Topics",
          "to": "/topics",
        },
      ]
    }
  >
    <NextSeo
      openGraph={
        Object {
          "description": "Find the topic you are planning to learn, or find a new topic to learn!",
          "title": "Browse Topics - Resource Rank",
        }
      }
      title="Topics"
    />
    <PageHeaderButton
      buttonContent="New"
      hideButton={true}
      linkTo="/topics/new"
      title="Topics"
      titleClassName="text-4xl"
    />
    <div
      className="mt-8 flex flex-col"
    >
      <input
        className="flex-grow form-control-lg"
        data-testid="SearchInput"
        onChange={[Function]}
        placeholder="Search topics..."
        type="text"
        value="search"
      />
      <div
        className="mt-8"
      >
        <LoadingErrorEmptyDataLayout
          dataDisplay={
            <TopicListing
              onPageClick={[MockFunction]}
              pageCount={NaN}
              selectedPage={1}
              topics={Array []}
            />
          }
          errorDisplay={
            <div
              className="text-center sm:text-left error-text"
            >
              Failed to load topics.
            </div>
          }
          hasData={false}
          hasError={false}
          isLoading={false}
          loadingDisplay={
            <div
              className="text-center"
            >
              <LoadingSpinner
                color="#292929"
                size={50}
              />
            </div>
          }
          noDataDisplay={
            <div
              className="text-center sm:text-left"
            >
              No topics have been created.
            </div>
          }
        />
      </div>
    </div>
  </BreadcrumbLayout>
</div>
`;
